server/plugins/messageStorage/sqlite.ts

```diff
@@ -91,43 +109,81 @@ class SqliteMessageStorage implements ISqliteMessageStorage {
 		}
 	}
 
-	async run_migrations() {
+	async setup_new_db() {
 		for (const stmt of schema) {
 			await this.serialize_run(stmt, []);
 		}
 
+		await this.serialize_run("INSERT INTO options (name, value) VALUES ('schema_version', ?)", [
+			currentSchemaVersion.toString(),
+		]);
+	}
+
+	async current_version(): Promise<number> {
+		const have_options = await this.serialize_get(
+			"select 1 from sqlite_master where type = 'table' and name = 'options'"
+		);
+
+		if (!have_options) {
+			return 0;
+		}
+
 		const version = await this.serialize_get(
 			"SELECT value FROM options WHERE name = 'schema_version'"
 		);
 
 		if (version === undefined) {
-			// new table
-			await this.serialize_run(
-				"INSERT INTO options (name, value) VALUES ('schema_version', ?)",
-				[currentSchemaVersion]
-			);
-			return;
+			// technically shouldn't happen, means something created a schema but didn't populate it
+			// we'll try our best to recover
+			return 0;
 		}
 
 		const storedSchemaVersion = parseInt(version.value, 10);
+		return storedSchemaVersion;
+	}
 
-		if (storedSchemaVersion === currentSchemaVersion) {
-			return;
-		}
-
-		if (storedSchemaVersion > currentSchemaVersion) {
-			throw `sqlite messages schema version is higher than expected (${storedSchemaVersion} > ${currentSchemaVersion}). Is The Lounge out of date?`;
-		}
-
+	async _run_migrations(dbVersion: number) {
 		log.info(
-			`sqlite messages schema version is out of date (${storedSchemaVersion} < ${currentSchemaVersion}). Running migrations if any.`
+			`sqlite messages schema version is out of date (${dbVersion} < ${currentSchemaVersion}). Running migrations.`
 		);
 
+		const to_execute = necessaryMigrations(dbVersion);
+
+		for (const stmt of to_execute.map((m) => m.stmts).flat()) {
+			await this.serialize_run(stmt, []);
+		}
+
 		await this.serialize_run("UPDATE options SET value = ? WHERE name = 'schema_version'", [
-			currentSchemaVersion,
+			currentSchemaVersion.toString(),
 		]);
 	}
 
+	async run_migrations() {
+		const version = await this.current_version();
+
+		if (version > currentSchemaVersion) {
+			throw `sqlite messages schema version is higher than expected (${version} > ${currentSchemaVersion}). Is The Lounge out of date?`;
```

Comment:
nah, I don't think that's a good idea.
We will add incompatible changes which won't work if we add features on top.

Why do you wanna support downgrades?

Reply:
I'd be ok with adding "down" scripts, but I sure as hell don't add code to run with more than one schema on a version. There should be one and only one state

Reply:
> We will add incompatible changes which won't work if we add features on top.

I don't think I understand the scenario you're worried about. Could you spell out a series of events to help me?

> Why do you wanna support downgrades?

Rolling back an upgrade is what I'm concerned about. A user upgrades, something's broken that they rely on, they rollback the upgrade to wait for the next release. With the functionality I highlighted, this user would still have a broken release after rolling back.

Reply:
I'd rather roll forward than backwards... 
If a user really wants to downgrade they can, they simply can start over with a new db.

It's simple really, anything but the most trivial migration *will* break old code, as the semantics differ.
Say I split networks / message type to their own table, as we are storing tons of strings for no reason where an int would do.
Old code has no way of interpreting that.

Even if you have "compatible" changes, sqlite by default doesn't enforce much, meaning old code is bound to enter state which new code would disallow / fixup.
Now you constantly have to write defensive code as you may have any corrupt state based on whatever the old table default is.

In any case, if you are dead set to support downgrades, fine. I'll defer to you, but then let's do downgrade migrations.
Is that a reasonable compromise?

Reply:
> Is that a reasonable compromise?

It's reasonable but it would also be very complex to execute. So I don't think it's what we should do.

Downgrade migrations aren't straightforward. The downgrade migration code would need to be in the previous version that you're rolling back to if we followed the strategy we're using for upgrade migrations. We'd probably want to use a different strategy for migrations instead if we wanted downgrade migrations.

I'm not really against doing them but since we don't currently have any migrations that need downgrading it seems like unnecessary complexity.

> It's simple really, anything but the most trivial migration will break old code, as the semantics differ.

I understand your concern now, thanks for spelling it out. This is what I meant to address by saying you should add a comment to the migrations list saying that our migration code is very simple and doesn't handle the case of migrations that aren't backwards compatible with previous versions.

This comment will (most likely) prevent us from breaking anything in a future migration accidentally.

--

Here's my argument for removing your throw:

* The throw is currently dead code. It won't do anything to prevent downgrades from the migration you added because the throw won't exist on the version the user is downgrading to.
* The only migration currently is backwards compatible.
* Quite complex migrations can be made to be backwards incompatible: it's not that lofty a goal.
* A comment in the migration list should be sufficient to prevent accidental breakages.

But I don't contribute much to this project currently so I don't actually have much stake and I'll defer to ya'll. I'll remove my blocking review.

Reply:
> I'll remove my blocking review.

Actually I have no idea how to do remove a request for changes without explicitly approving so I'll leave the state as-is.

Reply:
>The throw is currently dead code. It won't do anything to prevent downgrades from the migration you added because the throw won't exist on the version the user is downgrading to.
The only migration currently is backwards compatible.

Yes, that's by design 😉

The point is that we now add infrastructure that can in the future do stuff.
That's the reason I'm not actually doing anything with the new field.


>Quite complex migrations can be made to be backwards incompatible: it's not that lofty a goal.

Frankly, if you want to do anything at all you need to break backwards compatibility.
That's why you run migrations.


>A comment in the migration list should be sufficient to prevent accidental breakages.

But i want to break it, adding features on top that can't be done as is with the current schema is the point.

>It's reasonable but it would also be very complex to execute. So I don't think it's what we should do.

It's not really.
I can add a downmigration table that stores down migrations inside of the DB.
As the only thing we in the next release do is not used that's alright if a user downgrades now.
Anything in the future would be handled correctly unless they migrate to the version we have now.
But frankly why should they?
At most you'd downgrade one version if we really messed up. Which should not happen in the first place.

Reply:
> Frankly, if you want to do anything at all you need to break backwards compatibility.
> That's why you run migrations.

I don't want to push any further on the argument I've been making. But I do want to at least impart some cool knowledge: you really can make almost any migration backwards compatible. It can take some consideration, but in most cases it's really not bad.

Basically every migration in any even-somewhat-major website is backwards compatible, since you need to be able to rollback when the alternative is having an outage for awhile. I'm sure there's some cool blog posts about how to do it in various cases if you're interested. I bet ChatGPT could talk about it too 😂. That thing is pretty good at exploratory stuff like this.

Reply:
Then... how do you extract stuff to their own tables?
You can't in a compatible way.

And you want to, depending on your trade offs.

We know have a giant large table, it doesn't scale.
Even a 7-8GB DB crawls to a halt on certain operations say search (as we are doing a linear search with a very inefficient query)